File: framerail/src/lib/server/deepwell/watching.ts
import { client } from "$lib/server/deepwell"
import type { RequestContext } from "$lib/server/load/request-ctx"

export type WatchScope = "site" | "category" | "page"
export interface WatchSubscription {
  scope: WatchScope
  target_id: number
}
export interface WatchingPreferences {
  email_enabled: boolean
  auto_watch: boolean
}
export interface WatchingEvent {
  event_id: number
  page_id: number
  revision_id: number
  previous_revision_id: number | null
  title: string
  slug: string
  actor: string
  created_at: string
  event_type: "create" | "edit"
}
export interface WatchingActivity {
  items: WatchingEvent[]
  next_before: number | null
}

export interface WatchingChange extends WatchingEvent {
  before_text: string | null
  after_text: string
}

export function watchingPreferencesGet(
  context: RequestContext
): Promise<WatchingPreferences> {
  return client.request("watching_preferences_get", {}, context)
}

export function watchingPreferencesSet(
  preferences: WatchingPreferences,
  context: RequestContext
): Promise<WatchingPreferences> {
  return client.request("watching_preferences_set", preferences, context)
}

export function watchingSubscriptions(
  siteId: number,
  context: RequestContext
): Promise<WatchSubscription[]> {
  return client.request("watching_subscriptions", { site_id: siteId }, context)
}

export function watchingSubscriptionSet(
  siteId: number,
  scope: WatchScope,
  targetId: number,
  watching: boolean,
  context: RequestContext
): Promise<{ watching: boolean }> {
  return client.request(
    "watching_subscription_set",
    { site_id: siteId, scope, target_id: targetId, watching },
    context
  )
}

export function watchingActivity(
  siteId: number,
  beforeEventId: number | undefined,
  limit: number,
  context: RequestContext
): Promise<WatchingActivity> {
  return client.request(
    "watching_activity",
    {
      site_id: siteId,
      ...(beforeEventId === undefined ? {} : { before_event_id: beforeEventId }),
      limit
    },
    context
  )
}

export function watchingChange(
  eventId: number,
  context: RequestContext
): Promise<WatchingChange | null> {
  return client.request("watching_change", { event_id: eventId }, context)
}

export function watchingUnsubscribe(token: string): Promise<{ unsubscribed: boolean }> {
  return client.request("watching_unsubscribe", { token })
}

File: deepwell/src/endpoints/watching.rs
//! Watcher preferences, subscriptions, Activity, and token-authorized email opt-out.

use super::prelude::*;
use crate::services::watching::activity::{self, ActivityPage, ChangeDetail};
use crate::services::watching::subscriptions::{
    self, WatchPreferences, WatchScope, WatchSubscription,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
struct ActivityRequest {
    site_id: i64,
    before_event_id: Option<i64>,
    limit: Option<u64>,
}

#[derive(Debug, Deserialize)]
struct ChangeRequest {
    event_id: i64,
}

#[derive(Deserialize)]
struct UnsubscribeRequest {
    token: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct UnsubscribeOutput {
    unsubscribed: bool,
}

pub async fn watching_activity(
    ctx: &ServiceContext<'_>,
    params: Params<'static>,
) -> Result<ActivityPage> {
    let input: ActivityRequest = parse!(params, Request);
    activity::list_activity(
        ctx,
        input.site_id,
        input.before_event_id,
        input.limit.unwrap_or(20),
    )
    .await
}

pub async fn watching_change(
    ctx: &ServiceContext<'_>,
    params: Params<'static>,
) -> Result<Option<ChangeDetail>> {
    let input: ChangeRequest = parse!(params, Request);
    activity::get_change(ctx, input.event_id).await
}

pub async fn watching_unsubscribe(
    ctx: &ServiceContext<'_>,
    params: Params<'static>,
) -> Result<UnsubscribeOutput> {
    let input: UnsubscribeRequest = parse!(params, Request);
    Ok(UnsubscribeOutput {
        unsubscribed: activity::unsubscribe_email(ctx, &input.token).await?,
    })
}

#[derive(Debug, Deserialize)]
struct ListSubscriptions {
    site_id: i64,
}

#[derive(Debug, Deserialize)]
struct SetSubscription {
    site_id: i64,
    scope: WatchScope,
    target_id: i64,
    watching: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct SetSubscriptionOutput {
    watching: bool,
}

pub async fn watching_preferences_get(
    ctx: &ServiceContext<'_>,
    _params: Params<'static>,
) -> Result<WatchPreferences> {
    subscriptions::preferences_get(ctx).await
}

pub async fn watching_preferences_set(
    ctx: &ServiceContext<'_>,
    params: Params<'static>,
) -> Result<WatchPreferences> {
    let preferences: WatchPreferences = parse!(params, Request);
    subscriptions::preferences_set(ctx, preferences).await
}

pub async fn watching_subscriptions(
    ctx: &ServiceContext<'_>,
    params: Params<'static>,
) -> Result<Vec<WatchSubscription>> {
    let ListSubscriptions { site_id } = parse!(params, Request);
    subscriptions::subscriptions(ctx, site_id).await
}

pub async fn watching_subscription_set(
    ctx: &ServiceContext<'_>,
    params: Params<'static>,
) -> Result<SetSubscriptionOutput> {
    let SetSubscription {
        site_id,
        scope,
        target_id,
        watching,
    } = parse!(params, Request);
    let watching =
        subscriptions::subscription_set(ctx, site_id, scope, target_id, watching).await?;
    Ok(SetSubscriptionOutput { watching })
}
